langchain/src/vectorstores/typeorm.ts

```diff
@@ -0,0 +1,307 @@
+    postgresqlVectorStore.documentEntity = TypeORMDocumentEntity;
+
+    if (!postgresqlVectorStore.appDataSource.isInitialized) {
+      await postgresqlVectorStore.appDataSource.initialize();
+    }
+
+    postgresqlVectorStore._verbose =
+      typeof process !== "undefined"
+        ? // eslint-disable-next-line no-process-env
+          process.env?.LANGCHAIN_VERBOSE !== undefined
+        : false;
+
+    return postgresqlVectorStore;
+  }
+
+  async addDocuments(documents: Document[]): Promise<void> {
+    const texts = documents.map(({ pageContent }) => pageContent);
+    return this.addVectors(
+      await this.embeddings.embedDocuments(texts),
+      documents
+    );
+  }
+
+  async addVectors(vectors: number[][], documents: Document[]): Promise<void> {
+    const { createHash } = await import("node:crypto");
+    const rows = vectors.map((embedding, idx) => {
+      const embeddingString = `[${embedding.join(",")}]`;
+      const hash = createHash("sha1");
+      const documentRow = {
+        pageContent: documents[idx].pageContent,
+        embedding: embeddingString,
+        metadata: documents[idx].metadata,
+        id: documents[idx].id ?? getUniqueIDFromDocument(documents[idx]),
+        sourceType: getSourceTypeFromDocument(documents[idx]),
+        sourceName: getSourceNameFromDocument(documents[idx]),
+        hash:
+          documents[idx].hash ??
+          hash.update(documents[idx].pageContent).digest("hex"),
+      } as TypeORMVectorStoreDocument;
+
+      return documentRow;
+    });
+
+    const documentRepository = this.appDataSource.getRepository(
+      this.documentEntity
+    );
+
+    // For a given document, we delete all documents in the database that have not the same id as
+    // the ones we are trying to upsert because it means that the split of the file is not the same anymore
+    for (const doc of rows) {
+      if (!doc.sourceName || !doc.sourceType) {
+        continue;
+      }
+      // Search for existing document with sourceName and type
+      const documentsInDatabase = await documentRepository.find({
+        where: {
+          sourceName: doc.sourceName,
+          sourceType: doc.sourceType,
+        },
+      });
+
+      const documentsThatWillBeUpserted = rows.filter(
+        (d) => d.sourceName === doc.sourceName
+      );
+      const idsDocumentsThatWillBeUpserted = documentsThatWillBeUpserted.map(
+        (d) => d.id
+      );
+
+      const idsDocumentsThatWeShouldDeleted = documentsInDatabase
+        .filter((d) => !idsDocumentsThatWillBeUpserted.includes(d.id))
+        .map((d) => d.id);
+
+      this.verboseConsoleLog(
+        `Ids has changed for document ${doc.sourceName} and ${doc.sourceType}, action: delete this ids: ${idsDocumentsThatWeShouldDeleted}`
+      );
+
+      await documentRepository.delete({
+        id: {
+          in: idsDocumentsThatWeShouldDeleted,
+        },
+      });
+    }
+
+    const documentsToUpsert = [];
```

Comment:
we compare the hash of the document in the database with the one we want to insert to see if we need to recreate the embeddings. It's faster and cost less